src/main/java/pl/damianszczepanik/jenkins/buildhistorymanager/model/actions/DeleteArtifactsMatchingPatternsAction.java

```diff
@@ -0,0 +1,168 @@
+ */
+public class DeleteArtifactsMatchingPatternsAction extends Action {
+    private static final Logger LOG = Logger.getLogger(DeleteArtifactsMatchingPatternsAction.class.getName());
+
+    private String includePatterns;
+    private String excludePatterns;
+
+    @DataBoundConstructor
+    public DeleteArtifactsMatchingPatternsAction() {
+        // Jenkins stapler requires to have public constructor with @DataBoundConstructor
+    }
+
+    public String getIncludePatterns() {
+        return includePatterns;
+    }
+
+    @DataBoundSetter
+    public void setIncludePatterns(String includePatterns) {
+        this.includePatterns = includePatterns;
+    }
+
+    public String getExcludePatterns() {
+        return excludePatterns;
+    }
+
+    @DataBoundSetter
+    public void setExcludePatterns(String excludePatterns) {
+        this.excludePatterns = excludePatterns;
+    }
+
+    static boolean isDirectoryEmpty(Path path) throws IOException {
+        try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(path)) {
+            return !directoryStream.iterator().hasNext();
+        }
+    }
+
+    // if 'file' is on a different node, this FileCallable will be transferred to that node and executed there.
+    public static final class Delete implements FileCallable<Void> {
+        private static final long serialVersionUID = -7712623306942117400L;
+        private final File archiveRootFile;
+
+        public Delete(File archiveRootFile) {
+            this.archiveRootFile = archiveRootFile;
+        }
+
+        @Override 
+        public Void invoke(File file, VirtualChannel channel) throws IOException {
+            Set<File> directories = new HashSet<>();
+            deleteFileOrLogError(file, directories);
+            deleteEmptyDirectoriesAndParents(directories);
+            return null;
+        }
+
+        void deleteEmptyDirectoriesAndParents(Set<File> directories) throws IOException {
+            for (File dir : directories) {
+                if (shouldDeleteDirectory(dir)) {
+                    Util.deleteFile(dir);
+                    deleteParentDirectories(dir.getParentFile());
+                }
+            }
+        }
+
+        boolean shouldDeleteDirectory(File dir) throws IOException {
+            if (!isValidDirectory(dir)) {
+                return false;
+            }
+
+            File parent = dir.getParentFile();
+            return hasValidParent(parent);
+        }
+
+        boolean hasValidParent(File parent) {
+            return this.archiveRootFile.equals(parent);
+        }
+
+        boolean isValidDirectory(File dir) throws IOException {
+            return dir != null && dir.isDirectory() && isDirectoryEmpty(dir.toPath());
+        }
+
+        void deleteParentDirectories(File directory) {
+            File parent = directory;
+
+            while (shouldDelete(parent)) {
+                deleteDirectory(parent);
+                parent = parent.getParentFile();
```

Comment:
there is some risk here that you will delete whole storage up to root directory

Reply:
The `shouldDelete()` method makes sure we never go above the `archive` directory of a run.